src/Command/Ingester/Native/NativeUserIngesterCommand.php

```diff
@@ -178,58 +185,113 @@ protected function execute(InputInterface $input, OutputInterface $output): int
             $progressBar->setMaxSteps($numberOfUsers);
             $progressBar->start();
 
-            $index = $this->getAvailableStartIndex();
             $lineItemCollection = $this->fetchLineItems();
 
-            foreach ($source->getContent() as $row) {
+            $batchedRawUsers = [];
+            $numberOfProcessedRows = 1;
+            foreach ($source->getContent() as $rawUser) {
+                $batchedRawUsers[] = $rawUser;
+                if ($numberOfProcessedRows % $this->batchSize === 0) {
+                    $this->processBatchedRawUsers($batchedRawUsers, $lineItemCollection);
+                    $batchedRawUsers = [];
+                    $progressBar->advance($this->batchSize);
+                }
+
+                $numberOfProcessedRows++;
+            }
+
+            if ($batchedRawUsers) {
+                $this->processBatchedRawUsers($batchedRawUsers, $lineItemCollection);
+            }
+        } catch (Throwable $exception) {
+            $this->errors[] = $exception->getMessage();
+            $this->userQueryParts = [];
+            $this->assignmentQueryParts = [];
+        } finally {
+            $this->refreshSequences($resultSetMapping);
+            $progressBar->finish();
+        }
+
+        foreach ($this->errors as $error) {
+            $this->symfonyStyle->error($error);
+        }
+
+        return (int)!empty($this->errors);
+    }
+
+    /**
+     * @param LineItem[] $lineItemCollection
+     *
+     * @throws NonUniqueResultException
+     * @throws NoResultException
+     */
+    private function processBatchedRawUsers(array $rawUsers, array $lineItemCollection): void
+    {
+        $usernamesToCheck = array_map(
+            static function (array $rawUser) {
+                return $rawUser['username'];
+            },
+            $rawUsers
+        );
+
+        $existingUsernames = [];
+        foreach ($this->userRepository->findBy(['username' => array_unique($usernamesToCheck)]) as $user) {
+            $existingUsernames[$user->getId()] = $user->getUsername();
+        }
+
+        $userIndex = $this->getAvailableUserStartIndex();
+        $assignmentIndex = $this->getAvailableAssignmentStartIndex();
+
+        $resultSetMapping = new ResultSetMapping();
+        $user = new User();
+        foreach ($rawUsers as $rawUser) {
+            if (in_array($rawUser['username'], $existingUsernames, true)) {
+                $userId = array_search($rawUser['username'], $existingUsernames, true);
+
+                $this->assignmentQueryParts[] = sprintf(
```

Comment:
Possible memory leak issue. Are we cleaning these attributes from time to time?

Reply:
Yes we are clearing it around line 340 if I am not mistaken.